source/fuzz/transformation.h

```diff
@@ -0,0 +1,84 @@
+// limitations under the License.
+
+#ifndef SOURCE_FUZZ_TRANSFORMATION_H_
+#define SOURCE_FUZZ_TRANSFORMATION_H_
+
+#include "source/opt/ir_context.h"
+
+namespace spvtools {
+namespace fuzz {
+
+// The base class for all metamorphic transformations.
+//
+// Rules for transformations
+// -------------------------
+//
+// - Immutability: a transformation must be immutable.
+// - Ability to copy and serialize: to ensure that a copy of a transformation,
+//     possibly saved out to disk and read back again, is indistinguishable
+//     from the original transformation, thus a transformation must not depend
+//     only on well-defined pieces of state, such as instruction ids.  It must
+//     not rely on state such as pointers to instructions and blocks.
+// - Determinism: the effect of a transformation on a module be a deterministic
+//     function of the module and the transformation.  Any randomization should
+//     be applied before creating the transformation, not during its
+//     application.
+// - Well-defined and precondition: the 'IsApplicable' method should only
+//     return true if the transformation can be cleanly applied to the given
+//     module, to mutate it into a valid and semantically-equivalent module, as
+//     long as the module is initially valid.
+// - Ability to test precondition on any valid module: 'IsApplicable' should be
+//     designed so that it is safe to ask whether a transformation is
+//     applicable to an arbitrary valid module.  For example, if a
+//     transformation involves a block id, 'IsApplicable' should check whether
+//     the module indeed has a block with that id, and return false if not.  It
+//     must not assume that there is such a block.
+// - Documented precondition: while the implementation of 'IsApplicable' should
+//     should codify the precondition, the method should be commented in the
+//     header file for a transformation with a precise English description of
+//     the precondition.
+// - Documented effect: while the implementation of 'Apply' should codify the
+//     effect of the transformation, the method should be commented in the
+//     header file for a transformation with a precise English description of
+//     the effect.
+class Transformation {
+ public:
+  Transformation() = default;
+
+  virtual ~Transformation() = default;
+
+  // A precondition that determines whether the transformation can be cleanly
+  // applied to the SPIR-V module, such that semantics are preserved.
+  // Subclasses must document the precondition in their header file using
+  // precise English.
+  virtual bool IsApplicable(opt::IRContext* context) = 0;
+
+  // Requires that the transformation is applicable.  Applies the
+  // transformation, mutating the given SPIR-V module.
+  virtual void Apply(opt::IRContext* context) = 0;
+
+  // Returns true if and only if the module does not currently define the given
+  // id.
+  bool IsFreshId(opt::IRContext* context, uint32_t id);
```

Comment:
Should `IsFreshId` and `UpdateModuleIdBound` be here? Should they be static, protected, or maybe even global functions, or in some other "utilities" class? I feel like the `Transformation` class is otherwise a pure interface, which I like.

Reply:
I have opted for the global functions approach, creating a fuzzer_util.h/cpp pair.